sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,387 @@
+
+static void _override_node(const ipv6_addr_t *addr, unsigned iface,
+                           _nib_onl_entry_t *node);
+static inline bool _node_unreachable(_nib_onl_entry_t *node);
+
+void _nib_init(void)
+{
+#ifdef TEST_SUITES
+    _prime_def_router = NULL;
+    _next_removable.next = NULL;
+    memset(_nodes, 0, sizeof(_nodes));
+    memset(_def_routers, 0, sizeof(_def_routers));
+    memset(_nis, 0, sizeof(_nis));
+#endif
+    /* TODO: load ABR information from persistent memory */
+}
+
+_nib_onl_entry_t *_nib_onl_alloc(const ipv6_addr_t *addr, unsigned iface)
+{
+    _nib_onl_entry_t *node = NULL;
+
+    assert(addr != NULL);
+    DEBUG("nib: Allocating on-link node entry (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_NUMOF; i++) {
+        _nib_onl_entry_t *tmp = &_nodes[i];
+
+        if ((_nib_onl_get_if(tmp) == iface) &&
+            (ipv6_addr_equal(addr, &tmp->ipv6))) {
+            /* exact match */
+            DEBUG("  %p is an exact match\n", (void *)tmp);
+            return tmp;
+        }
+        if ((node == NULL) && (tmp->mode == _EMPTY)) {
+            node = tmp;
+        }
+    }
+    if (node != NULL) {
+        DEBUG("  using %p\n", (void *)node);
+        _override_node(addr, iface, node);
+    }
+#if ENABLE_DEBUG
+    else {
+        DEBUG("  NIB full\n");
+    }
+#endif
+    return node;
+}
+
+static inline bool _is_gc(_nib_onl_entry_t *node)
+{
+    return ((node->mode & ~(_NC)) == 0) &&
+           ((node->info & GNRC_IPV6_NIB_NC_INFO_AR_STATE_MASK) ==
+            GNRC_IPV6_NIB_NC_INFO_AR_STATE_GC);
+}
+
+static inline _nib_onl_entry_t *_cache_out_onl_entry(const ipv6_addr_t *addr,
+                                                     unsigned iface,
+                                                     uint16_t cstate)
+{
+    /* Use clist as FIFO for caching */
+    _nib_onl_entry_t *first = (_nib_onl_entry_t *)clist_lpop(&_next_removable);
+    _nib_onl_entry_t *tmp = first, *res = NULL;
+
+    DEBUG("nib: Searching for replaceable entries (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    if (tmp != NULL) {
+        do {
+            if (_is_gc(tmp)) {
+                DEBUG("nib: Removing neighbor cache entry (addr = %s, "
+                      "iface = %u) ",
+                      ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                       sizeof(addr_str)),
+                      _nib_onl_get_if(tmp));
+                DEBUG("for (addr = %s, iface = %u)\n",
+                      ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)),
+                      iface);
+                res = tmp;
+                res->mode = _EMPTY;
+                _override_node(addr, iface, res);
+                /* masked above already */
+                res->info = cstate;
+                res->mode = _NC;
+            }
+            else {
```

Comment:
`else` is not needed here and introduces more cpu cycles to re-check `tmp` in the next while iteration. Consider to move the content of this `else` block outside.